oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/R2dbcReactiveOAuth2AuthorizedClientService.java

```diff
@@ -0,0 +1,323 @@
+			OAuth2AuthorizedClientHolder oAuth2AuthorizedClientHolder) {
+		return this.clientRegistrationRepository
+				.findByRegistrationId(oAuth2AuthorizedClientHolder.getClientRegistrationId())
+				.switchIfEmpty(Mono
+						.error(dataRetrievalFailureException(oAuth2AuthorizedClientHolder.getClientRegistrationId())))
+				.map(clientRegistration -> new OAuth2AuthorizedClient(clientRegistration,
+						oAuth2AuthorizedClientHolder.getPrincipalName(),
+						oAuth2AuthorizedClientHolder.getoAuth2AccessToken(),
+						oAuth2AuthorizedClientHolder.getRefreshToken()));
+	}
+
+	private Mono<Integer> updateAuthorizedClient(OAuth2AuthorizedClient authorizedClient, Authentication principal) {
+		GenericExecuteSpec executeSpec = databaseClient.sql(UPDATE_AUTHORIZED_CLIENT_SQL);
+		for (Entry<String, Parameter> entry : this.authorizedClientParametersMapper.apply(authorizedClient, principal)
+				.entrySet()) {
+			executeSpec = executeSpec.bind(entry.getKey(), entry.getValue());
+		}
+		return executeSpec.fetch().rowsUpdated();
+	}
+
+	private Mono<Integer> insertAuthorizedClient(OAuth2AuthorizedClient authorizedClient, Authentication principal) {
+		GenericExecuteSpec executeSpec = databaseClient.sql(SAVE_AUTHORIZED_CLIENT_SQL);
+		for (Entry<String, Parameter> entry : this.authorizedClientParametersMapper.apply(authorizedClient, principal)
+				.entrySet()) {
+			executeSpec = executeSpec.bind(entry.getKey(), entry.getValue());
+		}
+		return executeSpec.fetch().rowsUpdated();
+	}
+
+	private Throwable dataRetrievalFailureException(String clientRegistrationId) {
+		return new DataRetrievalFailureException("The ClientRegistration with id '" + clientRegistrationId
+				+ "' exists in the data source, " + "however, it was not found in the ClientRegistrationRepository.");
+	}
+
+	private static final class OAuth2AuthorizedClientParametersMapper
+			implements BiFunction<OAuth2AuthorizedClient, Authentication, Map<String, Parameter>> {
+
+		@Override
+		public Map<String, Parameter> apply(OAuth2AuthorizedClient authorizedClient, Authentication principal) {
+
+			final Map<String, Parameter> parameters = new HashMap<>();
+
+			final ClientRegistration clientRegistration = authorizedClient.getClientRegistration();
+			final OAuth2AccessToken accessToken = authorizedClient.getAccessToken();
+			final OAuth2RefreshToken refreshToken = authorizedClient.getRefreshToken();
+
+			parameters.put("clientRegistrationId", Parameter.from(clientRegistration.getRegistrationId()));
+			parameters.put("principalName", Parameter.from(principal.getName()));
+			parameters.put("accessTokenType", Parameter.from(accessToken.getTokenType().getValue()));
+			parameters.put("accessTokenValue",
+					Parameter.from(ByteBuffer.wrap(accessToken.getTokenValue().getBytes(StandardCharsets.UTF_8))));
+			parameters.put("accessTokenIssuedAt",
+					Parameter.from(LocalDateTime.ofInstant(accessToken.getIssuedAt(), ZoneOffset.UTC)));
```

Comment:
I wonder if we should use `Timestamp.from()` same as in `JdbcOAuth2AuthorizedClientService`?

Reply:
Timestamp is not a valid datatype. Based on https://r2dbc.io/spec/0.8.3.RELEASE/spec/html/, sql type TIMESTAMP should be mapped as ```java.time.LocalDateTime``` if it represents a date and time without a timezone.